stdlib/experimental/dynamic/dynamic_test.flux

```diff
@@ -10,3 +11,56 @@ testcase dynamic_not_comparable {
         want: /unsupported/,
     )
 }
+
+testcase dynamic_does_not_rewrap {
+    a = dynamic.dynamic(v: 123)
+    b = dynamic.dynamic(v: a)
+
+    testing.assertEqualValues(want: true, got: dynamic._equal(a, b))
+}
+
+// asArray should blow up if you feed it a dynamic that doesn't have an array in it.
+testcase asArray_errors_on_nonarray {
+    testing.shouldError(
+        fn: () => dynamic.dynamic(v: 123) |> dynamic.asArray(),
+        want: /unable to convert/,
+    )
+}
+
+// Verify we can pass an array of dynamic elements into dynamic() to wrap it, then unwrap it with asArray.
+testcase asArray_accepts_actual_array {
+    arr = dynamic.dynamic(v: [dynamic.dynamic(v: 123)]) |> dynamic.asArray()
+
+    testing.assertEqualValues(
+        want: true,
+        got: dynamic._equal(a: dynamic.dynamic(v: 123), b: arr[0]),
+    )
+}
+
+// This is similar to the "actual array" test except that the elements in the
+// array are not wrapped in dynamic ahead of time. asArray therefore needs to
+// wrap the elements prior to producing the `[dynamic]` it'll return.
```

Comment:
I would expect it to be a bug for `dynamic.jsonParse` a dynamic array whose elements weren't also wrapped. Maybe `asArray` should assert in this case?

On the other hand, I can also imagine other sources of dynamic data that, unlike JSON, are able to guarantee homogeneity of arrays. In that case, _maybe_ it might make sense for `asArray` to leave the elements themselves unwrapped and produce a `[int]` or whatever?

I guess what I'm saying is that it seems like auto-wrapping in `asArray` could cover up bugs elsewhere in the stdlib (if we're being strict about wrapping), or it could be making things needlessly difficult (if we're not). Thoughts?

Reply:
Thinking some more about this, the signature of `asArray` is `(<-v: dynamic) => [dynamic]` so it _needs_ to produce a `[dynamic]`. I think `asArray` should produce an error rather than wrap---but still not 100%.

Reply:
I've been mulling this over - the problem of who's responsibility is is to wrap. 

From my side I was trying to weigh whether or not the wrapping should happen in the initial `dynamic()` call when it receives an `[int]`. I opted to do the check on the `toArray` side to defer the extra work of wrapping each element until it would be needed.

If we say that `toArray` returns `[A]` instead of `[dynamic]` the extra wrapping could be skipped entirely for the homogeneous case but we sacrifice enforcement of the static types matching reality, I think.

The one thing I can say is if the caller sends in an `[int]` I would like for them to not need to manually wrap each element themselves in the script. That would be punitive.

Reply:
> It seems like you're suggesting that we can think of dynamic(v: ...) as producing a dynamic value where any collections also have their values wrapped with dynamic, but that as an implementation detail we are not wrapping them. We are wrapping them lazily upon access as an optimization.

This is precisely what I was thinking. Clearly there are situations where we can't afford to be lazy, specifically the `json.parse` which needs to establish the element type for arrays it produces as it builds them.

I think I agree that symmetry is desirable and so I think that means adding a recursive structure to `dynamic(v: ...)`. Perhaps there should be some code share between this and the json parser since I imagine the job is effectively the same in both cases.

If this sounds right, I can relocate my naive array wrapping code into `dynamic(v:...)` as a stop-gap and we can coalesce once both our branches have landed.

> I'm wondering about the use case of dynamic(v: ...). Do we ever expect it to be used outside of testing?

The original use case I identified was in building json bodies to send out via http POST.

Reply:
It sounds right to me. We can change things later on if we need, since I imagine all this will be in `experimental` for a while.